src/backend/distributed/executor/partitioned_intermediate_results.c

```diff
@@ -0,0 +1,515 @@
+															   partitionColumnIndex);
+	Var *partitionColumn = makeVar(partitionColumnIndex, partitionColumnIndex,
+								   partitionColumnAttr->atttypid,
+								   partitionColumnAttr->atttypmod,
+								   partitionColumnAttr->attcollation, 0);
+	DistTableCacheEntry *shardSearchInfo =
+		QueryTupleShardSearchInfo(minValuesArray, maxValuesArray,
+								  partitionMethod, partitionColumn);
+
+	/* prepare the output destination */
+	EState *estate = CreateExecutorState();
+	MemoryContext tupleContext = GetPerTupleMemoryContext(estate);
+	PartitionedResultDestReceiver *dest =
+		CreatePartitionedResultDestReceiver(resultIdPrefixString, partitionColumnIndex,
+											partitionCount, tupleDescriptor, binaryCopy,
+											shardSearchInfo, tupleContext);
+
+	/* execute the query */
+	PortalRun(portal, FETCH_ALL, false, true, (DestReceiver *) dest,
+			  (DestReceiver *) dest, NULL);
+
+	/* construct the output result */
+	TupleDesc returnTupleDesc = NULL;
+	Tuplestorestate *tupleStore = SetupTuplestore(fcinfo, &returnTupleDesc);
+	resultInfo->returnMode = SFRM_Materialize;
+	resultInfo->setResult = tupleStore;
+	resultInfo->setDesc = returnTupleDesc;
+
+	for (int partitionIndex = 0; partitionIndex < partitionCount; partitionIndex++)
+	{
+		uint64 recordsWritten = 0;
+		uint64 bytesWritten = 0;
+		Datum values[3];
+		bool nulls[3];
+
+		if (dest->partitionDestReceivers[partitionIndex] != NULL)
+		{
+			FileDestReceiverStats(dest->partitionDestReceivers[partitionIndex],
+								  &recordsWritten, &bytesWritten);
+		}
+
+		memset(values, 0, sizeof(values));
+		memset(nulls, 0, sizeof(nulls));
+
+		values[0] = Int32GetDatum(partitionIndex);
+		values[1] = UInt64GetDatum(recordsWritten);
+		values[2] = UInt64GetDatum(bytesWritten);
+
+		tuplestore_putvalues(tupleStore, returnTupleDesc, values, nulls);
+	}
+
+	tuplestore_donestoring(tupleStore);
+	PortalDrop(portal, false);
+	FreeExecutorState(estate);
+
+	PG_RETURN_INT64(1);
+}
+
+
+/*
+ * StartPortalForQueryExecution creates and starts a portal which can be
+ * used for running the given query.
+ */
+static Portal
+StartPortalForQueryExecution(const char *queryString)
+{
+	Query *query = ParseQueryString(queryString, NULL, 0);
+
+	int cursorOptions = CURSOR_OPT_PARALLEL_OK;
+	PlannedStmt *queryPlan = pg_plan_query(query, cursorOptions, NULL);
+
+	Portal portal = CreateNewPortal();
+
+	/* don't display the portal in pg_cursors, it is for internal use only */
+	portal->visible = false;
+
+	PortalDefineQuery(portal, NULL, queryString, "SELECT", list_make1(queryPlan), NULL);
+	int eflags = 0;
+	PortalStart(portal, NULL, eflags, GetActiveSnapshot());
+
+	return portal;
+}
+
+
+/*
+ * QueryTupleShardSearchInfo returns a DistTableCacheEntry which has enough
+ * information so that FindShardInterval() can find the shard corresponding
+ * to a tuple.
+ */
+static DistTableCacheEntry *
+QueryTupleShardSearchInfo(ArrayType *minValuesArray, ArrayType *maxValuesArray,
```

Comment:
I think this might take about a day. I can do it before code-freeze if everything else is done by then. Created an issue: https://github.com/citusdata/citus/issues/3341